lib/platform/linux-uart.c

```diff
@@ -0,0 +1,432 @@
+					      "%02x", *(uint8_t *)(src + n - 1 - i));
+			b += snprintf(b, sizeof(cmd), " 0x%x\r", crc);
+		break;
+		case SWITCHTEC_UART_GAS_READ:
+			b += snprintf(cmd, sizeof(cmd),
+				      "gasrd -c -s 0x%x %zu\r", addr, n);
+		break;
+	}
+
+	ret = write(fd, cmd, strlen(cmd));
+	if (ret != strlen(cmd))
+		return -1;
+	else
+		return 0;
+}
+
+static int read_resp_line(int fd, char *str)
+{
+	int ret;
+	int i;
+	int cnt = 0;
+
+	for (i = 0; i < RETRY_NUM; i++) {
+		ret = read(fd, str + cnt, sizeof(str));
+		if (ret < 0)
+			return ret;
+	
+		cnt += ret;
+		str[cnt] = '\0';
+	
+		if (strrchr(str, ':') + 5 == strrchr(str, '>'))
+			return 0;
+	}
+	
+	return -1;
+}
+
+static int cli_control(struct switchtec_dev *dev, const char *str)
+{
+	int ret;
+	char rtn[1024];
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+
+	ret = send_cmd(udev->fd, str, SWITCHTEC_UART_CONTROL, 0, 0, 0, 0);
+	if (ret)
+		return ret;
+
+	ret =  read_resp_line(udev->fd, rtn);
+	if (ret)
+		return ret;
+		
+	return 0;
+}
+
+#ifdef __CHECKER__
+#define __force __attribute__((force))
+#else
+#define __force
+#endif
+
+static void uart_close(struct switchtec_dev *dev)
+{
+	struct switchtec_uart *udev =  to_switchtec_uart(dev);
+	cli_control(dev, echo_on);
+
+	if (dev->gas_map)
+		munmap((void __force *)dev->gas_map, dev->gas_map_size);
+
+	flock(udev->fd, LOCK_UN);
+	close(udev->fd);
+	free(udev);
+}
+
+static int map_gas(struct switchtec_dev *dev)
+{
+	void *addr;
+	dev->gas_map_size = 4 << 20;
+
+	addr = mmap(NULL, dev->gas_map_size, PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (addr == MAP_FAILED)
+		return -1;
+
+	dev->gas_map = (gasptr_t __force)addr;
+
+	return 0;
+}
+
+#undef __force
+
+static gasptr_t uart_gas_map(struct switchtec_dev *dev, int writeable, size_t *map_size)
+{
+	if (map_size)
+		*map_size = dev->gas_map_size;
+
+	return dev->gas_map;
+}
+
+static void uart_memcpy_from_gas(struct switchtec_dev *dev, void *dest, const void __gas *src, size_t n)
+{
+	int ret;
+	int raddr, rnum, rcrc;
+	int i, j;
+	char *pos;
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+	uint32_t addr = (uint32_t)(src - (void __gas *)dev->gas_map);
+	char gas_rd_rtn[102400];
+	uint8_t buf[n+sizeof(addr)];
+	uint8_t *ptr = buf;
+	uint8_t cal;
+
+	for (i = 0; i< RETRY_NUM; i++) {
+		ret =  send_cmd(udev->fd, NULL, SWITCHTEC_UART_GAS_READ, addr, n, src, 0);
+		if (ret)
+			raise(SIGBUS);
+
+		ret = read_resp_line(udev->fd, gas_rd_rtn);
+		if (ret)
+			raise(SIGBUS);
+
+		if(sscanf(gas_rd_rtn, "%*[^<]<0x%x> [%d Byte]%*[^:]: 0x%x%*[^:]:",
+				  &raddr, &rnum, &rcrc) != 3)
+			raise(SIGBUS);
+
+		if ((raddr != addr) || (rnum != n))
+			raise(SIGBUS);
+
+		addr = htobe32(addr);
+		memcpy(ptr, &addr, sizeof(addr));
+		ptr += sizeof(addr);
+
+		pos = strchr(gas_rd_rtn, ']');
+		pos += 2;
+		for (j = 0; j < n; j++)
+			*ptr++ = strtol(pos, &pos, 16);
+
+		cal = crc8(buf, sizeof(addr) + n, 0, true);
+		if (cal == rcrc){
+			memcpy(dest, buf + sizeof(addr), n);
```

Comment:
Why don't we just put the data directly into _dest_ instead of allocating another buffer and copying it?

You can always do two crc calls such as:

```c
crc = crc8(&addr, sizeof(addr), 0, true)
crc = crc8(dest, n, crc, false);
```

Reply:
thanks, crc8 have this feature for calculation based on the last result.